backend/models/loginRecord.js

```diff
@@ -0,0 +1,127 @@
+/**
+ *	Copyright (C) 2014 3D Repo Ltd
+ *
+ *	This program is free software: you can redistribute it and/or modify
+ *	it under the terms of the GNU Affero General Public License as
+ *	published by the Free Software Foundation, either version 3 of the
+ *	License, or (at your option) any later version.
+ *
+ *	This program is distributed in the hope that it will be useful,
+ *	but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *	GNU Affero General Public License for more details.
+ *
+ *	You should have received a copy of the GNU Affero General Public License
+ *	along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+ "use strict";
+ const db = require("../handler/db");
+ 
+ //detects edge as browser but not device
+ var uaParserJs = require('ua-parser-js');
+ var device = require('device');
+ var ip2location = require('ip-to-location');
+
+ const LoginRecord = {};
+ 
+ LoginRecord.saveLoginRecord = async (req) => {
+        var loginRecord = {
+                _id : req.sessionID,
+                loginTime : new Date(),
+                ipAddr: req.ips[0] || req.ip                    
+        }
+
+        const referrer = req.header('Referer');
+        if(referrer != null){
+                loginRecord.referrer = referrer;
+        }
+
+        const location = await getLocationFromIPAddress(loginRecord.ipAddr);   
+        loginRecord.location = {
+                country: location.country_name,
+                city: location.city
+        };    
+       
+         const userAgentString = req.headers['user-agent'];
+         var uaInfo = {};
+
+         if(isUserAgentFromPlugin(userAgentString)){
+                uaInfo = getUserAgentInfoFromPlugin(userAgentString);
+         }
+         else{
+                uaInfo = getUserAgentInfoFromBrowser(userAgentString);
+         }
+
+         loginRecord.application = uaInfo.application;
+         loginRecord.engine = uaInfo.engine;
+         loginRecord.os = uaInfo.os;
+         loginRecord.device = uaInfo.device;          
+
+         const col = await db.getCollection("loginRecords", req.body.username);
+         await col.insertOne(loginRecord);
+ }
+ 
+ //Format: 
+ //PLUGIN: {OS Name}/{OS Version} {Host Software Name}/{Host Software Version} {Plugin Type}/{Plugin Version}
+ //Example:
+ //PLUGIN: Windows/10.0.19042.0 REVIT/2021.1 PUBLISH/4.15.0
+ function getUserAgentInfoFromPlugin(userAgentString){
```

Comment:
declare functions as arrow functions
```suggestion
const getUserAgentInfoFromPlugin => (userAgentString) {
```

There's some nuance difference between this and classic function definition, main benefit is the binding of `this` - the `this` keyword is super confusing in the JavaScript world...